sympy/simplify/fu.py

```diff
@@ -1020,6 +1020,48 @@ def f(rv):
 
     return bottom_up(rv, f)
 
+def _TR11(rv):
+    """
+    Helper for TR11 to find the `base` argument of TR11.
+
+    Examples
+    ========
+
+    >>> from sympy.simplify.fu import _TR11 as T
+    >>> from sympy import cos, sin, pi
+    >>> from sympy.abc import x
+    >>> T(sin(x/3)*sin(2*x)*sin(x/4)/(cos(x/6)*cos(x/8)))
+    4*sin(x/8)*sin(x/6)*sin(2*x)
+    >>> T(sin(x/3)/cos(x/6))
+    2*sin(x/6)
+    >>> T(sin(2*x)*cos(x/8)/sin(x/4))
+    sin(2*x)/(2*sin(x/8))
+
+    """
+    def f(rv):
+        arg_denoms = dict()
+        for fi in rv.args:
+            if fi.is_Pow and fi.exp.is_negative and fi.base.func in (sin, cos):
+                arg_denoms[fi.base.args[0]] = fi.base.func
```

Comment:
you might want to make `arg_denoms` = defaultdict(set)` and here do `arg_denoms[fi.base.args[0]].add(fi.base.func)` since a given arg might be contained in `sin` and `cos`. Then below instead of popping the arg, just pop the function from the function set in the dictionary's value for that key.

Reply:
That will surely increasing efficiency while popping the function. But then, wouldn't I have to check whether the set of functions for each arg is empty or not in elif before further calculations? Or, does there exist a faster way to check for that too?

Reply:
As in `if not (fi.func == cos and arg_denoms[arg_denom] == cos) :` becoming `if not (fi.func == cos and cos in arg_denoms[arg_denom]): `?

Reply:
I was referring to `if len(arg_denoms):` in line 1048. I guess this would become `if any(len(arg_denoms[arg]) != 0 for args in arg_denoms)`.

Reply:
Actually, I don't think you need this guard. If there is nothing to iterate over it will be done (after doing nothing) in a short amount of time.